File: PRACTICA3/QytetetRuby/lib/titulo_propiedad.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module ModeloQytetet
  class TituloPropiedad
    
    attr_reader :nombre,:precio_compra,:alquiler_base,:factor_revalorizacion,:hipoteca_base,:precio_edificar,:hipotecada,:num_casas,:num_hoteles
    attr_writer :hipotecada
    
    def initialize(nombre,precio_compra,alquiler_base,factor_revalorizacion,hipoteca_base,precio_edificar)
      @nombre = nombre
      @precio_compra = precio_compra
      @alquiler_base = alquiler_base
      @factor_revalorizacion = factor_revalorizacion
      @hipoteca_base = hipoteca_base
      @precio_edificar = precio_edificar
      @hipotecada = false
      @num_casas = 0
      @num_hoteles = 0
      @propietario = nil
    end
   
    def calcular_coste_cancelar
    
        coste_cancelar = calcular_coste_hipotecar + (0.1*calcular_coste_hipotecar)
        return coste_cancelar
    end
    
    def calcular_coste_hipotecar
    
        coste_hipoteca = (@hipoteca_base + @num_casas *0.5*@hipoteca_base + @num_hoteles * @hipoteca_base)
        return coste_hipoteca
    end
    
    def calcular_importe_alquiler
    
        coste_alquiler = @alquiler_base + (@num_casas*5 + @num_hoteles*2)
        return coste_alquiler
    end
    
    def calcular_precio_venta
    
         precio_venta =  (@precio_compra + (@num_casas + @num_hoteles) * @precio_edificar * @factor_revalorizacion)
        return precio_venta
    end
    
    def cancelar_hipoteca
    
    end    
    
    
     def cobrar_alquiler(coste)
    
     end
    
    
    def edificar_casa
      @num_casas = @num_casas +1
    end   
    
    
    def edificar_hotel
      @num_hoteles = @num_hoteles + 1
    end   
    
    
    def to_s
      "\nNombre: #{@nombre} \n Precio_compra: #{@precio_compra} \n Alquiler_base: #{@alquiler_base} \n Factor_revalorizacion: #{@factor_revalorizacion} \n Hipoteca_base: #{@hipoteca_base} \n Precio_edificar: #{@precio_edificar} \n Hipotecada: #{@hipotecada} \n Num_casas: #{@num_casas} \n Num_hoteles: #{@num_hoteles}"
    end
  end
end

File: PRACTICA3/QytetetJava/src/modeloqytetet/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author jesus
 */
public class TituloPropiedad {
    private String nombre;
    private boolean hipotecada;
    private int precioCompra;
    private int alquilerBase;
    private float factorRevalorizacion;
    private int hipotecaBase;
    private int precioEdificar;
    private int numCasas;
    private int numHoteles;
    private Jugador propietario;

    public TituloPropiedad(String nombre, int precioCompra, int alquilerBase, float factorRevalorizacion, int hipotecaBase, int precioEdificar) {
        this.nombre = nombre;
        this.hipotecada = false;
        this.precioCompra = precioCompra;
        this.alquilerBase = alquilerBase;
        this.factorRevalorizacion = factorRevalorizacion;
        this.hipotecaBase = hipotecaBase;
        this.precioEdificar = precioEdificar;
        this.hipotecada = false;
        this.numCasas = 0;
        this.numHoteles = 0;
    }
    
    int calcularCosteCancelar()
    {
        int costeCancelar = (int) (calcularCosteHipotecar() + (0.1*calcularCosteHipotecar()));
        return costeCancelar;
    }
    
    int calcularCosteHipotecar()
    {
        int costeHipoteca = (int) (hipotecaBase + numCasas * 0.5 * hipotecaBase +  numHoteles * hipotecaBase);
        return costeHipoteca;
    }
    
    int calcularImporteAlquiler()
    {
       int costeAlquiler = alquilerBase + (numCasas*5 + numHoteles*2);
       return costeAlquiler;
    }
    
    int calcularPrecioVenta()
    {
        int precioVenta = (int) (precioCompra + (numCasas + numHoteles) * precioEdificar * factorRevalorizacion);
        return precioVenta;
    }
    
    void cancelarHipoteca()
    {
        hipotecada = false;
    }
    
    void cobrarAlquiler(int coste)
    {
        
    }
    
    void edificarCasa()
    {
        numCasas+=1;
    }
    
    void edificarHotel()
    {
        numHoteles+=1;
    }

    public String getNombre() {
        return nombre;
    }

    public boolean isHipotecada() {
        return hipotecada;
    }

    public int getPrecioCompra() {
        return precioCompra;
    }

    public int getAlquilerBase() {
        return alquilerBase;
    }

    public float getFactorRevalorizacion() {
        return factorRevalorizacion;
    }

    public int getHipotecaBase() {
        return hipotecaBase;
    }

    public int getPrecioEdificar() {
        return precioEdificar;
    }

    public int getNumCasas() {
        return numCasas;
    }

    public int getNumHoteles() {
        return numHoteles;
    }

    public void setHipotecada(boolean hipotecada) {
        this.hipotecada = hipotecada;
    }
    
    int hipotecar()
    {
        setHipotecada(true);
        int costeHipoteca = calcularCosteHipotecar();
        return costeHipoteca;
    }
    
    int pagarAlquiler()
    {
        int costeAlquiler = calcularImporteAlquiler();
        propietario.modificarSaldo(costeAlquiler);
        return costeAlquiler;
    }
    
    boolean propietarioEncarcelado()
    {
        return propietario.isEncarcelado();
    }
    
    boolean tengoPropietario()
    {
        return (propietario!=null);
    }

    Jugador getPropietario() {
        return propietario;
    }

    void setPropietario(Jugador propietario) {
        this.propietario = propietario;
    }
    
    
    
    

    @Override
    public String toString() {
        return "TituloPropiedad{" + "nombre=" + nombre + ", hipotecada=" + hipotecada + ", precioCompra=" + precioCompra + ", alquilerBase=" + alquilerBase + ", factorRevalorizacion=" + factorRevalorizacion + ", hipotecaBase=" + hipotecaBase + ", precioEdificar=" + precioEdificar + ", numCasas=" + numCasas + ", numHoteles=" + numHoteles + '}' + "\n";
    }
    
    
    
    
}
